dbms/src/Flash/Mpp/GRPCSendQueue.h

```diff
@@ -0,0 +1,285 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Common/MPMCQueue.h>
+#include <common/logger_useful.h>
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+#endif
+#include <grpcpp/grpcpp.h>
+#ifdef __clang__
+#pragma clang diagnostic pop
+#endif
+
+#include <functional>
+
+namespace DB
+{
+
+namespace tests
+{
+class TestGRPCSendQueue;
+} // namespace tests
+
+/// In grpc cpp framework, the tag that is pushed into grpc completion
+/// queue must be inherited from `CompletionQueueTag`.
+class KickTag : public ::grpc::internal::CompletionQueueTag
+{
+public:
+    enum class Status
+    {
+        // No tag.
+        NONE,
+        // Waiting for kicking.
+        WAITING,
+        // Queuing in the grpc completion queue.
+        QUEUING,
+    };
+
+    KickTag(std::mutex * m, const LoggerPtr l)
+        : mu(m)
+        , log(l)
+    {}
+
+    ~KickTag() override
+    {
+        std::unique_lock lock(*mu);
+
+        RUNTIME_ASSERT(status == Status::NONE, log, "status {} is not none", status);
+    }
+
+    bool FinalizeResult(void ** tag_, bool * /*status*/) override
+    {
+        std::unique_lock lock(*mu);
+
+        RUNTIME_ASSERT(status == Status::QUEUING, log, "status {} is not queuing", status);
+        status = Status::NONE;
+
+        *tag_ = tag;
+        tag = nullptr;
+
+        return true;
+    }
+
+    Status getStatus()
+    {
+        std::unique_lock lock(*mu);
+        return status;
+    }
+
+    /// Should acquire lock.
+    bool toWaiting(void * new_tag)
+    {
+        RUNTIME_ASSERT(new_tag != nullptr, log, "new_tag is nullptr");
+
+        if (status != Status::NONE)
+        {
+            return false;
+        }
+        status = KickTag::Status::WAITING;
+        tag = new_tag;
+        return true;
+    }
+
+    /// Should acquire lock.
+    bool toQueuing()
+    {
+        if (status != Status::WAITING)
+        {
+            return false;
+        }
+        RUNTIME_ASSERT(tag != nullptr, log, "status is waiting but tag is nullptr");
+        status = Status::QUEUING;
+        return true;
+    }
+
+private:
+    friend class tests::TestGRPCSendQueue;
+
+    std::mutex * mu;
```

Comment:
Current usages to mu is error-prone in that:
1. not all methods lock mu, actually `toQueuing` implicitly requires the caller has locked mu.
2. It' not clear who holds mu.

Reply:
Changed `toQueuing` and `toWaiting` to private functions.